Sources/ServiceDiscovery/ServiceDiscovery.swift

```diff
@@ -12,120 +12,42 @@
 //
 //===----------------------------------------------------------------------===//
 
-import Atomics
-import Dispatch
-
 // MARK: - Service discovery protocol
 
 /// Provides service instances lookup.
 ///
 /// ### Threading
 ///
 /// `ServiceDiscovery` implementations **MUST be thread-safe**.
-public protocol ServiceDiscovery: AnyObject {
+public protocol ServiceDiscovery {
     /// Service identity type
-    associatedtype Service: Hashable
+    associatedtype Service
     /// Service instance type
-    associatedtype Instance: Hashable
-
-    /// Default timeout for lookup.
-    var defaultLookupTimeout: DispatchTimeInterval { get }
+    associatedtype Instance
 
-    /// Performs a lookup for the given service's instances. The result will be sent to `callback`.
+    /// Performs async lookup for the given service's instances.
     ///
     /// ``defaultLookupTimeout`` will be used to compute `deadline` in case one is not specified.
     ///
-    /// ### Threading
-    ///
-    /// `callback` may be invoked on arbitrary threads, as determined by implementation.
-    ///
     /// - Parameters:
     ///   - service: The service to lookup
-    ///   - deadline: Lookup is considered to have timed out if it does not complete by this time
-    ///   - callback: The closure to receive lookup result
-    func lookup(_ service: Service, deadline: DispatchTime?, callback: @escaping (Result<[Instance], Error>) -> Void)
+    ///
+    /// -  Returns: A listing of service instances.
+    func lookup(_ service: Service, deadline: ContinuousClock.Instant?) async throws -> [Instance]
```

Comment:
I think returning `Set<Instance>` is more appropriate

Reply:
addressed